datafusion-cli/src/exec.rs

```diff
@@ -165,9 +166,15 @@ pub async fn exec_from_repl(
             }
             Ok(line) => {
                 rl.add_history_entry(line.trim_end())?;
-                match exec_and_print(ctx, print_options, line).await {
-                    Ok(_) => {}
-                    Err(err) => eprintln!("{err}"),
+                tokio::select! {
+                    res = exec_and_print(ctx, print_options, line) => match res {
```

Comment:
Looks like CTRL+C just getting out of the loop without actual query interruption?

So the machine resources is not released when the user tries to run next query after CTRL+C?

Reply:
My understanding was that when control arrives there, tasks were already interrupted. But I could be wrong -- in that case this PR needs to address your point

Reply:
I think this code is correct.

My understanding of  what happens is that if the signal is handled, it will `drop` the future running `exec_and_print` which will implicitly  is `drop`ed which should then propagate back and cancel outstanding tasks 

There are a variety of tests that validate that outstanding tasks are cancelled when their containing streams are dropped such as https://github.com/apache/arrow-datafusion/blob/1179a76567892b259c88f08243ee01f05c4c3d5c/datafusion/physical-plan/src/coalesce_partitions.rs#L216-L234 

If we find instances where dropping a stream doesn't cancel the ongoing execution, I think we should file that as a bug and fix it

Reply:
🤔  it might be good to add some documentation to the https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#tymethod.execute to explain the drop / cancel behavior and expectations this better. I will add that to my list

Reply:
I think this is correct, too

Per [tokio::select documentation](https://docs.rs/tokio/latest/tokio/macro.select.html):

> Waits on multiple concurrent branches, returning when the first branch completes, **cancelling the remaining branches.**

This is also what I tried initially: https://github.com/apache/arrow-datafusion/issues/1279#issuecomment-1316660929

But I had the roadblock of `exec_and_print` being blocking at that time. That doesn't seem to be the case anymore so this should be good.

Reply:
Here is a proposed PR with some clarifying documentation : https://github.com/apache/arrow-datafusion/pull/8747

Reply:
> But I had the roadblock of exec_and_print being blocking at that time. That doesn't seem to be the case anymore so this should be good.

I think @berkaysynnada  / @ozankabak  made the output mostly streaming in https://github.com/apache/arrow-datafusion/pull/8651